### ISSUE TITLE
Very strange web dialog - is it possible/feasible to implement a workaround in NVDA for this scenario?

### ISSUE THREAD

**Issue by Palacee-hun** [nvaccess/nvda#11534]
### Is your feature request related to a problem? Please describe.
There's an online shopping website for food delivery quite popular in Hungary. I've been using it a lot for many years now and it has been especially important in the Covid lockdown period.
Recently the website got a quite thorough redesign. As often is the case, with the redesign came some accessibility issues (the website used to have almost none of these). The most critical is a really strange specimen.
I can add items to my order basket by pressing the "Add" button underneath the given item. Quite standard, no problems there. But there are certain items (and indeed they are frequent) for which there are additional options to choose from (they have a wide variety). Sometimes it's mandatory to make a choice before putting into basket, sometimes not.
If I add such an item with options to my basket, a very strange web dialog comes up. The most particular thing about that is that it can't be read in browse mode, unlike almost all other web dialogs I've encountered so far can. Whatever I do in browse mode, NVDA just says "blank". In interactive (focus) mode, I can tab around, but here comes another nasty "surprise": except two controls NVDA doesn't read any names for the radio buttons/checkboxes present in the dialog (which would represent my choices). I don't recall when I have seen such nameless web controls last - it was ages ago. And the third particularity is that NVDA reports this strange dialog as invisible! It's possible to exit this dialog by pressing Ctrl-NVDA-space.
This behaviour is for my Chrome-derived browser (called Iridium). In Firefox NVDA doesn't even tell me about this particular dialog. It simply behaves as if nothing has happened after pressing the "Add"button, so I wouldn't have had the chance to figure all this out there.
Although this website has an English version, it would be very tough to explain how to get to such an item to bring up this nasty web dialog, so instead I opt for providing developer info (NVDA-F1) for three relevant objects.
Here it is for the dialog itself:
```
INFO - globalCommands.GlobalCommands.script_navigatorObject_devInfo (11:27:34.298) - MainThread (4072):
Developer info for navigator object:
name: None
role: ROLE_DIALOG
roleText: None
states: STATE_INVISIBLE, STATE_FOCUSABLE, STATE_FOCUSED
isFocusable: True
hasFocus: True
Python object: <NVDAObjects.Dynamic_WebDialogDocumentIa2WebIAccessible object at 0x03ED7DB0>
Python class mro: (<class 'NVDAObjects.Dynamic_WebDialogDocumentIa2WebIAccessible'>, <class 'NVDAObjects.behaviors.WebDialog'>, <class 'NVDAObjects.IAccessible.chromium.Document'>, <class 'NVDAObjects.IAccessible.ia2Web.Document'>, <class 'NVDAObjects.IAccessible.ia2Web.Ia2Web'>, <class 'NVDAObjects.IAccessible.IAccessible'>, <class 'NVDAObjects.window.Window'>, <class 'NVDAObjects.NVDAObject'>, <class 'documentBase.TextContainerObject'>, <class 'baseObject.ScriptableObject'>, <class 'baseObject.AutoPropertyObject'>, <class 'object'>)
description: None
location: RectLTWH(left=8, top=84, width=541, height=15)
value: None
appModule: <'appModuleHandler' (appName 'iridium', process ID 4552) at address 493ba70>
appModule.productName: 'Iridium'
appModule.productVersion: '2019.04.73.0'
TextInfo: <class 'NVDAObjects.IAccessible.IA2TextTextInfo'>
windowHandle: 8782820
windowClassName: 'Chrome_RenderWidgetHostHWND'
windowControlID: 367298136
windowStyle: 1445986304
extendedWindowStyle: 32
windowThreadID: 2960
windowText: 'Chrome Legacy Window'
displayText: ''
IAccessibleObject: <POINTER(IAccessible2) ptr=0x4f416d4 at 50360d0>
IAccessibleChildID: 0
IAccessible event parameters: windowHandle=8782820, objectID=-4, childID=-912799
IAccessible accName: None
IAccessible accRole: ROLE_SYSTEM_DIALOG
IAccessible accState: STATE_SYSTEM_FOCUSED, STATE_SYSTEM_INVISIBLE, STATE_SYSTEM_FOCUSABLE, STATE_SYSTEM_VALID (1081348)
IAccessible accDescription: None
IAccessible accValue: ''
IAccessible2 windowHandle: 8782820
IAccessible2 uniqueID: -912799
IAccessible2 role: ROLE_SYSTEM_DIALOG
IAccessible2 states: IA2_STATE_OPAQUE (1024)
IAccessible2 attributes: 'display:block;tag:div;xml-roles:dialog;hidden:true;class:modal fade item-modifier-modal  in;id:choices-toppings-modal;'
```
And for a nameless option checkbox:
```
INFO - globalCommands.GlobalCommands.script_navigatorObject_devInfo (11:30:20.546) - MainThread (4072):
Developer info for navigator object:
name: None
role: ROLE_CHECKBOX
roleText: None
states: STATE_INVISIBLE, STATE_FOCUSABLE, STATE_FOCUSED
isFocusable: True
hasFocus: True
Python object: <NVDAObjects.IAccessible.ia2Web.Ia2Web object at 0x050310F0>
Python class mro: (<class 'NVDAObjects.IAccessible.ia2Web.Ia2Web'>, <class 'NVDAObjects.IAccessible.IAccessible'>, <class 'NVDAObjects.window.Window'>, <class 'NVDAObjects.NVDAObject'>, <class 'documentBase.TextContainerObject'>, <class 'baseObject.ScriptableObject'>, <class 'baseObject.AutoPropertyObject'>, <class 'object'>)
description: None
location: RectLTWH(left=255, top=77, width=13, height=14)
value: ''
appModule: <'appModuleHandler' (appName 'iridium', process ID 4552) at address 493ba70>
appModule.productName: 'Iridium'
appModule.productVersion: '2019.04.73.0'
TextInfo: <class 'NVDAObjects.IAccessible.IA2TextTextInfo'>
windowHandle: 8782820
windowClassName: 'Chrome_RenderWidgetHostHWND'
windowControlID: 367298136
windowStyle: 1445986304
extendedWindowStyle: 32
windowThreadID: 2960
windowText: 'Chrome Legacy Window'
displayText: ''
IAccessibleObject: <POINTER(IAccessible2) ptr=0x4f89314 at 50360d0>
IAccessibleChildID: 0
IAccessible event parameters: windowHandle=8782820, objectID=-4, childID=-912800
IAccessible accName: None
IAccessible accRole: ROLE_SYSTEM_CHECKBUTTON
IAccessible accState: STATE_SYSTEM_FOCUSED, STATE_SYSTEM_INVISIBLE, STATE_SYSTEM_FOCUSABLE, STATE_SYSTEM_VALID (1081348)
IAccessible accDescription: None
IAccessible accValue: ''
IAccessible2 windowHandle: 8782820
IAccessible2 uniqueID: -912800
IAccessible2 role: ROLE_SYSTEM_CHECKBUTTON
IAccessible2 states: IA2_STATE_OPAQUE, IA2_STATE_CHECKABLE (263168)
IAccessible2 attributes: 'display:block;tag:input;checkable:true;hidden:true;'
```
And finally for a regular button in this dialog (fortunately it's the "put in basket"button, so I can work around all this for items for which choosing any options is not mandatory):
```
INFO - globalCommands.GlobalCommands.script_navigatorObject_devInfo (11:25:15.972) - MainThread (4072):
Developer info for navigator object:
name: 'TEDD A KOSÁRBA'
role: ROLE_BUTTON
roleText: None
states: STATE_INVISIBLE, STATE_OFFSCREEN, STATE_FOCUSED, STATE_FOCUSABLE
isFocusable: True
hasFocus: True
Python object: <NVDAObjects.Dynamic_ButtonIa2WebIAccessible object at 0x044A80B0>
Python class mro: (<class 'NVDAObjects.Dynamic_ButtonIa2WebIAccessible'>, <class 'NVDAObjects.IAccessible.Button'>, <class 'NVDAObjects.IAccessible.ia2Web.Ia2Web'>, <class 'NVDAObjects.IAccessible.IAccessible'>, <class 'NVDAObjects.window.Window'>, <class 'NVDAObjects.NVDAObject'>, <class 'documentBase.TextContainerObject'>, <class 'baseObject.ScriptableObject'>, <class 'baseObject.AutoPropertyObject'>, <class 'object'>)
description: None
location: RectLTWH(left=132, top=35, width=384, height=48)
value: ''
appModule: <'appModuleHandler' (appName 'iridium', process ID 4552) at address 493ba70>
appModule.productName: 'Iridium'
appModule.productVersion: '2019.04.73.0'
TextInfo: <class 'NVDAObjects.IAccessible.IA2TextTextInfo'>
windowHandle: 8782820
windowClassName: 'Chrome_RenderWidgetHostHWND'
windowControlID: 367298136
windowStyle: 1445986304
extendedWindowStyle: 32
windowThreadID: 2960
windowText: 'Chrome Legacy Window'
displayText: ''
IAccessibleObject: <POINTER(IAccessible2) ptr=0x4f41e24 at 5049e40>
IAccessibleChildID: 0
IAccessible event parameters: windowHandle=8782820, objectID=-4, childID=-912809
IAccessible accName: 'TEDD A KOSÁRBA'
IAccessible accRole: ROLE_SYSTEM_PUSHBUTTON
IAccessible accState: STATE_SYSTEM_FOCUSED, STATE_SYSTEM_INVISIBLE, STATE_SYSTEM_OFFSCREEN, STATE_SYSTEM_FOCUSABLE, STATE_SYSTEM_VALID (1146884)
IAccessible accDescription: None
IAccessible accValue: ''
IAccessible2 windowHandle: 8782820
IAccessible2 uniqueID: -912809
IAccessible2 role: ROLE_SYSTEM_PUSHBUTTON
IAccessible2 states: IA2_STATE_OPAQUE (1024)
IAccessible2 attributes: 'display:block;tag:button;xml-roles:button;hidden:true;class:product-add-to-cart-button js-toppings-add-to-cart button full;'
```

### Describe the solution you'd like
Upon encountering all this messy stuff I knew at once that some weird and very non-standard web coding nightmare lies behind this extraordinary dialog. So I told them about this. Fortunately they do care about screen reader accessibility and admitted that they didn't integrate accessibility awareness and testing during the redesign (quite common practice even for developers who care about this, "we'll do it later" approach). They promised to improve on this as soon as possible with the help of visually impaired testers.
But as no progress was made with this till now, I began to contemplate whether anything can be done about this on screen reader level. I am afraid several months (or even more) would pass before they get this right.
I am not particularly into web programming and not at all into IA2 or the very intricate cooperation between NVDA and web browsers. So I ask:
- is IA2 implementation is in play here behind this strange rendering of dialog content? - personally I think so, because Firefox and Chrome do quite different things with it
- is it possible and feasible to work around this in NVDA somehow?
- how could this anomaly be fixed on web code level? - if I could give a hint about this to web developers, that could speed up the fix.

**Comment by Adriani90** [comment]
It seems the web author has applie display:none or visibility:hidden or something like that on the dialog and its controls. I guess this is somehow visible for sighted people but not for the screen reader. At least this is what the state invisible in the developer info suggests.
@feerrenrut since we don't have a test case here, I am tending to think that this is an invalid issue and is rather a bad web authoring. What do you think?
@Palacee-hun did you write some feedback to the developers of the website?
Please also try to test with Google Chrome or with Microsoft Edge Chromium, in case you are on Windows 10 2004 update.

**Comment by Palacee-hun** [comment]
Yes, I did contact the web authors. I quote the relevant portion of my description:
"Upon encountering all this messy stuff I knew at once that some weird and very non-standard web coding nightmare lies behind this extraordinary dialog. So I told them about this. Fortunately they do care about screen reader accessibility and admitted that they didn't integrate accessibility awareness and testing during the redesign (quite common practice even for developers who care about this, "we'll do it later" approach). They promised to improve on this as soon as possible with the help of visually impaired testers.
But as no progress was made with this till now, I began to contemplate whether anything can be done about this on screen reader level. I am afraid several months (or even more) would pass before they get this right."
Maybe it wasn't clear that "they" refer to the web authors.
Basicly I came here with this issue because based on a lot of my experience with contacting web authors I am afraid it won't be fixed soon, and the web authors maybe clueless how to get this right. As I mentioned this is an important site, and will be even more if another Covid lockdown of some form arrives (the second wave of the pandemic has just started in my country).
I use Windows 7 ultimate 32-bit, so no MS Edge here.

**Comment by Palacee-hun** [comment]
If there is someone from Hungary here and willing to test and comment on this here, then I'll give "steps to reproduce" instructions. Otherwise it's pointless to provide such, as though the website has an English version, you need registration and a Hungarian street address to get to a "test case". The name of the site is www.netpincer.hu.

**Comment by Palacee-hun** [comment]
Recently I've got to know that MS Edge can indeed be installed on Win7, so I did so and tested this issue with it. The most up-to-date Edge (version 86.0.622.48) behaved exactly like I described for Iridium browser. That is no surprise by the way as both are Chromium-based. Updating to NVDA 2020.3 didn't make any difference in connection with this issue either.
Regarding web developers the only "improvement" they made after I had told them about this issue was that now trying to put ANY item in basket brings up this strange type of dialog, no matter what the product is. I can only use the website for anything after this change just because of a very lucky thing: the "put in basket" button in this dialog is still labelled and read out in interactive mode, so I can tab to it and press just that without selecting anything else or doing anything besides that in this dialog.
In my view this move of the web developers shows clearly that they are completely in the dark about this issue and haven't got any clue how to address this. I consider it probable that they just picked some form of a code template of this strange web dialog from the miriads of ready-made web control code templates, and they like it for a reason or another and they just stick to it. Maybe they don't have the legal rights to change the dialog template for some reason, and the causes of inaccessibility lie in that template.
All in all I don't see much hope that the web developers would fix this in the months to come. This strengthens my opinion I've formed for a long time now that screen readers are to correct the accessibility mistakes of web authors as much as possible. In this case If I had any insight into the extremely complex browse mode backend code of NVDA (yes, I tried to familiarise myself with it, but sorry it's unbelievably intricate), then as a start I would try to enumerate the cases when NVDA cannot render a web dialog in browse mode, as this is the heart of this issue.

**Comment by Palacee-hun** [comment]
A rare success story: somewhere in May out of the blue this strange and very inaccessible web dialog was transformed into a totally ordinary and thus entirely usable and accessible web dialog. I've just thanked the web author for this and sincerely hope that it stays so.
